handler/src/main/java/io/netty/handler/ssl/ReleasableOpenSslEngine.java

```diff
@@ -0,0 +1,1988 @@
+                        bytesConsumed, bytesProduced);
+            }
+
+            // Check to see if we received a close_notify message from the peer.
+            if (!receivedShutdown && (SSL.getShutdown(ssl) & SSL.SSL_RECEIVED_SHUTDOWN) == SSL.SSL_RECEIVED_SHUTDOWN) {
+                closeAll();
+            }
+
+            return newResult(bytesConsumed, bytesProduced, status);
+        }
+    }
+
+    private SSLEngineResult sslReadErrorResult(int err, int bytesConsumed, int bytesProduced) throws SSLException {
+        String errStr = SSL.getErrorString(err);
+
+        // Check if we have a pending handshakeException and if so see if we need to consume all pending data from the
+        // BIO first or can just shutdown and throw it now.
+        // This is needed so we ensure close_notify etc is correctly send to the remote peer.
+        // See https://github.com/netty/netty/issues/3900
+        if (SSL.pendingWrittenBytesInBIO(networkBIO) > 0) {
+            if (handshakeException == null && handshakeState != HandshakeState.FINISHED) {
+                // we seems to have data left that needs to be transfered and so the user needs
+                // call wrap(...). Store the error so we can pick it up later.
+                handshakeException = new SSLHandshakeException(errStr);
+            }
+            return new SSLEngineResult(OK, NEED_WRAP, bytesConsumed, bytesProduced);
+        }
+        throw shutdownWithError("SSL_read", errStr);
+    }
+
+    private int pendingAppData() {
+        // There won't be any application data until we're done handshaking.
+        // We first check handshakeFinished to eliminate the overhead of extra JNI call if possible.
+        return handshakeState == HandshakeState.FINISHED ? SSL.pendingReadableBytesInSSL(ssl) : 0;
+    }
+
+    private SSLEngineResult newResult(
+            int bytesConsumed, int bytesProduced, SSLEngineResult.HandshakeStatus status) throws SSLException {
+        return new SSLEngineResult(
+                getEngineStatus(), mayFinishHandshake(status != FINISHED ? getHandshakeStatus() : status)
+                , bytesConsumed, bytesProduced);
+    }
+
+    private void closeAll() throws SSLException {
+        receivedShutdown = true;
+        closeOutbound();
+        closeInbound();
+    }
+
+    private void rejectRemoteInitiatedRenegation() throws SSLHandshakeException {
+        if (rejectRemoteInitiatedRenegation && SSL.getHandshakeCount(ssl) > 1) {
+            // TODO: In future versions me may also want to send a fatal_alert to the client and so notify it
+            // that the renegotiation failed.
+            shutdown();
+            throw new SSLHandshakeException("remote-initiated renegotation not allowed");
+        }
+    }
+
+    public final SSLEngineResult unwrap(final ByteBuffer[] srcs, final ByteBuffer[] dsts) throws SSLException {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)